### PR TITLE
Make sure custom alpha param does not change 'none' colors in a list of colors

### LINKED ISSUE #27839: [Bug]: PathCollection using alpha ignores 'none' facecolors
### Bug summary

As title, using alpha parameter in PathCollection, either as a list or as a scalar value, causes 'none' facecolors to be ignored / filled

### Code for reproduction

```Python
import matplotlib.pyplot as plt
from matplotlib.collections import PathCollection
from matplotlib.path import Path

# https://matplotlib.org/stable/gallery/shapes_and_collections/compound_path.html

paths = [
    Path(
        [(1, 1), (1, 2), (2, 2), (2, 1), (0, 0)],
        [Path.MOVETO] + [Path.LINETO] * 3 + [Path.CLOSEPOLY]
    ),
    Path(
        [(4, 4), (5, 5), (5, 4), (0, 0)],
        [Path.MOVETO] + [Path.LINETO] * 2 + [Path.CLOSEPOLY]
    ),
]

fig, ax = plt.subplots()
collection = PathCollection(
    paths,
    edgecolors='black',
    facecolors=['blue', 'none'],
    alpha=[1, 1],
    # alpha=1
)
collection.set_transform(ax.transData)
ax.add_artist(collection)
ax.set(xlim=(0, 5), ylim=(0, 5))
plt.show()
```

### Actual outcome

actual

### Expected outcome

expected

### Additional information

_No response_

### Operating system

Windows

### Matplotlib Version

3.8.3

### Matplotlib Backend

_No response_

### Python version

3.12

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by chaoyihu**
## PR summary

This change suggests a fix and hopefully closes #27839 by making sure that 'none' colors in a list won't be affected by custom alpha param.

Before the fix, this code:
```
# colors::to_rgba_array()
509 if alpha is not None:
510     rgba[:, 3] = alpha
```
will change rgba array of color list ['blue', 'none'] from:
```
[[0. 0. 1. 1.]       # this is blue
 [0. 0. 0. 0.]]      # this has no color
```
to
```
[[0. 0. 1. 1.]       # this is still blue
 [0. 0. 0. 1.]]      # this becomes black:
```

After the fix, any 'none' color in a list of colors will be kept colorless.

A test is added to make sure the function works as expected.

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/tests/test_colors.py [hunk 1]
```diff
@@ -1660,3 +1663,25 @@ def test_set_cmap_mismatched_name():
     cmap_returned = plt.get_cmap("wrong-cmap")
     assert cmap_returned == cmap
     assert cmap_returned.name == "wrong-cmap"
+
+
+def test_pathcollection_alpha_none_facecolor():
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_colors.py:1668]: Please run the test on `to_rgba_array()` directly, not on `PathCollection`. While you've experienced the problem in `PatchCollection`, the actual fault and behavior change is in that function. It's better to keep the scope of unit tests very focussed.

Comment (chaoyihu) [reply lib/matplotlib/tests/test_colors.py]: Sure, thanks. I've rewritten the unit test to test specifically the changed code in `to_rgba_array`. Please let me know if there are any problems.
Plus, I am seeing building check failures due to subprocess timeouts which does not seem to be related to my code change. How should I fix this?

File: lib/matplotlib/tests/test_colors.py [hunk 2]
```diff
@@ -1660,3 +1661,11 @@ def test_set_cmap_mismatched_name():
     cmap_returned = plt.get_cmap("wrong-cmap")
     assert cmap_returned == cmap
     assert cmap_returned.name == "wrong-cmap"
+
+
+def test_to_rgba_array_none_color_with_alpha_param():
+    c = ["blue", "none"]
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_colors.py:1667]:
```suggestion
    # effective alpha for color "none" must always be 0 to achieve a vanishing color
    # even explicit alpha must be ignored
    c = ["blue", "none"]
```

File: lib/matplotlib/colors.py [hunk 3]
```diff
@@ -506,6 +506,9 @@ def to_rgba_array(c, alpha=None):
 
     if alpha is not None:
         rgba[:, 3] = alpha
+    if isinstance(c, Sequence):
+        rgba[:, 3] = [0 if isinstance(cc, str) and cc == 'none'
+                      else rgba[i, 3] for i, cc in enumerate(c)]
```
Comment (timhoffm) [inline lib/matplotlib/colors.py:511]: AFAICT, we can move this in: `to_rgba("none")` returns the correct alpha. Only the explicit alpha can mess this  up. Thus, we don't have to spend the extra effort for cases without explicit alpha.

Also using a numpy mask should be somewhat faster and clearer.

```suggestion
        if isinstance(c, Sequence):
            # ensure that an explicit alpha does not overwrite full transparency for "none"
            none_mask = np.array([isinstance(cc, str) and cc == "none" for cc in c])
            rgba[:, 3][none_mask] = 0
```

**Comment by timhoffm** [comment]
Something went wrong with the most recent commit it contains the plot_date deprecation, which is already on main. I suggest you squash and rebase on main.

**Comment by chaoyihu** [comment]
Sorry, I think it's because I synced fork by mistake. I made a new commit and the code changes are correct now. Also squashed the commits as suggested.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
CI failure is unrelated. I'm milestoning as 3.9 because I'd like to see this in there and the fix is easy to review.

File: lib/matplotlib/colors.py [hunk 4]
```diff
@@ -506,6 +506,11 @@ def to_rgba_array(c, alpha=None):
 
     if alpha is not None:
         rgba[:, 3] = alpha
+        if isinstance(c, Sequence):
+            # ensure that an explicit alpha does not overwrite full transparency
+            # for "none"
+            none_mask = np.array([isinstance(cc, str) and cc == "none" for cc in c])
```
Comment (QuLogic) [inline lib/matplotlib/colors.py:512]: I don't think this needs to be an array, does it?
```suggestion
            none_mask = [cbook._str_equal(cc, "none") for cc in c]
```

Comment (timhoffm) [reply lib/matplotlib/colors.py]: Yes. I took the liberty to commit. Please squash-merge.

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @chaoyihu! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**